strsep.3
.\" $Id: strsep.3,v 1.1 1998/10/13 07:12:09 garbled Exp $
.\" Copyright (c) 1990, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Chris Torek.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     from: @(#)strsep.3	8.1 (Berkeley) 6/9/93
.\"	$NetBSD: strsep.3,v 1.7 1998/02/05 18:50:55 perry Exp $
.\"
.Dd June 9, 1993
.Dt STRSEP 3
.Os
.Sh NAME
.Nm strsep
.Nd separate strings
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.Fd #include <string.h>
.Ft char *
.Fn strsep "char **stringp" "char *delim"
.Sh DESCRIPTION
The
.Fn strsep
function locates, in the null-terminated string referenced by
.Fa *stringp ,
the first occurrence of any character in the string
.Fa delim
(or the terminating
.Ql \e0
character) and replaces it with a
.Ql \e0 .
The location of the next character after the delimiter character
(or NULL, if the end of the string was reached) is stored in
.Fa *stringp .
The original value of
.Fa *stringp
is returned.
.Pp
An ``empty'' field, i.e. one caused by two adjacent delimiter characters,
can be detected by comparing the location referenced by the pointer returned
in
.Fa *stringp
to
.Ql \e0 .
.Pp
If
.Fa *stringp
is initially
.Dv NULL ,
.Fn strsep
returns
.Dv NULL .
.Sh EXAMPLES
The following uses
.Fn strsep
to parse a string, containing tokens delimited by white space, into an
argument vector:
.Bd -literal -offset indent
char **ap, *argv[10], *inputstring;

for (ap = argv; (*ap = strsep(&inputstring, " \et")) != NULL;)
	if (**ap != '\e0')
		++ap;
.Ed
.Sh HISTORY
The
.Fn strsep
function
is intended as a replacement for the
.Fn strtok
function.
While the
.Fn strtok
function should be preferred for portability reasons (it conforms to
.St -ansiC )
it is unable to handle empty fields, i.e. detect fields delimited by
two adjacent delimiter characters, or to be used for more than a single
string at a time.
The
.Fn strsep
function first appeared in
.Bx 4.4 .


CHANGES
$Id: CHANGES,v 1.15 2005/12/13 08:04:25 garbled Exp $

Add a -v option to all programs to tell the user what version of
ClusterIt they are running.  CVS will allways have _BETA tagged on the end,
and it will be removed at release.

Add RVT_CMD_ARGS to dvt, to allow passing of args to rvt

Incorporate patches from debian xvt package into rvt:    
Fix xvt manpage typos
Add -rv (reverse-video) option 
Better colormap handling
All other patches were either irrelevant (makefile fixes, etc) or allready
incorporated.

Fix pdf to work on linux hosts

Replace most instances of sprintf with snprintf. Did not touch rvt, as I would
prefer to keep that as unmodified as possible, and most of them would not
be easy to fix anyhow.

Replace most instances of malloc with calloc.   Mostly to prevent any
unanticipated wierdness, and a few of the mallocs were immediately followed
by memset commands anyhow.

Add RCMD_CMD_ARGS to most of the programs. This makes it easier to do
things like run "ssh -4" as your RCMD_CMD.

Stop allocating the command to be run on the remote end as buf[MAXBUF],
instead compute it at runtime and dynamically allocate it to prevent
possible overflow on a huge command.

Fix a bug noted by Jot Powers, where trailing whitespace on a GROUP or LUMP
entry would cause dsh -g to not match it.

Fix a bug noted by John Bales  where linux machines often recieved truncated 
output from the child ssh/rsh process.

Fix bug where all the programs would mangle argv[0] for ps.

* Release 2.3.1 *

Fix a bug noted by James F Wilkus on OpenBSD where dsh would segv when
reading the clusterfile.

Convert seq over to rseq fully, not just for linux anymore.  Sigh.

Fix the clusterit.spec file to work properly.  Tested on redhat

* Release 2.3 *

Made all remote commands use poll instead of a while loop to read the incoming
stream of data from stdout and stderr.  This fixes bug where linux ssh will
hang up.

Fixed linux command line argument processing.

Changed the -l <username> processing to invoke RCMD_CMD with user@host rather
than -l user host.  This makes it much more reliable.

Added an option to dsh to test remote ports before connecting.  This can be
useful where a remote host might be down, and a long timeout would occur
waiting for a response.

Updated rvt to use a newer set of sources from xvt.  Fixes questionable
licensing in xvt, and makes it more reliable.

Converted makefiles to autoconf for ease of porting.  No more stupid patch
sets which were a terrible burden to maintain and generate.

ANSIfyed most of the code (other than rvt). Nobody uses KNF anymore.. sigh..

Incorporated many patches sent by users from the internet.  I have
unfortunately lost track of what patch came from who, but thier help is
greatly appreciated.  Most of these patches either fixed bugs, added stability
or fixed porting bugs.  The xvt conversion was based mainly on a patch sent
in by a user.  My thanks go out to those of you who have sent them in.

# changes since 2.2

* Release 2.2 *
Missing this data, sorry
* Release 2.0 *

Added dvt and rvt..  an interactive form of dsh with xterms.

Fixed a bug in jsd and dsh where it would lock up running ssh waiting for
input.

Add a jsd dir, and move jsd.c from dsh to the jsd dir.  Add nearly complete
jsh program, and jcommon.[ch].

Add new common subdir and move a tonnage of crap in there.  Mainly
common.c, common.h and strsep.c.

Fix horrid bug in dsh where dsh | more would lock terminals up.  Still
need to fix in seq, run, probably pcp.

Add <netinet/if_ether.h> to barrierd.c to shut AIX C up.

Lots of minor bugfixes, mostly dealing with failures to open files, opening
0 byte clusters, etc etc..

Pcp was still using it's own code to parse strings and groups.  Fully rewrote
the back-end internals to use the new linked-list stuff, and utilize common.c.
Pcp's code is now virtually identical to dsh, whereas before it had diverged
significantly.

Error handling has finally graduated to where I no longer feel that it
"needs work".

Fix a place where malloced data was strcatted to without being bzeroed.

Group, exclusion and -w nodes are now dynamicly generated arrays.  Removes
the MAX_GROUPS restrictions.

The nodelist and grouplists are no longer maintained in any way similar to
the original setup.  Nodes are not stored in a flat array of nodes, with a
corresponding flat array of group names.  Nodes are now stores in a linked
list of machines to run on.  Nodes which are not to be run on, are not
included in this list, so the back-end routines only need to execute code
on every node they are handed in the linked list.  This is all handled in
common.c now.  This removes the all compile time restrictions on MAXNODES,
allowing the program to scale much more efficiently.

# changes since 1.3
* Release 1.3 *

Break dsh, run and seq's common code into a common.c and common.h file,
so I don't forget to add options to run and seq when fixing dsh.  Also, fix
some locations where run and seq were out of sync slightly.

Fix a bug in pdf where if the -m option was used, filesystems at 100%
would not be displayed.

Handle ^C's correctly.  If a remote process is hung, and it is the current
process being read, you may ^C it, and the interrupt will be sent to the
remote process.  This would be just like hitting ^C on the remote end, 
and will only affect the process you are currently reading from.

Added ability to pcp multiple files and wildcards to remote hosts.

Made pcp and dsh compile on an ANSI compiler (IBM C, SUN C)

Make output from dsh look nicer, and format well.

# changes since 1.2
* Release 1.2 *

Bug fixes in dsh to deal with the start of a fangroup, and
a race condition, where large output would cause it to hang
waiting for descriptors to die.

Fix a bug in dsh that caused it to lose data on AIX, and 
dump core when transitioning fangroups.

Added the rungroup, and exclusion commands to all the programs.

Fix pdf to work on a heterogeneous cluster.


common/strsep.c
/* $Id: strsep.c,v 1.2 1998/10/13 07:05:23 garbled Exp $ */
/*	$NetBSD: strsep.c,v 1.7 1998/02/03 18:49:23 perry Exp $	*/

/*-
 * Copyright (c) 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
#if 0
static char sccsid[] = "@(#)strsep.c	8.1 (Berkeley) 6/4/93";
#else
__RCSID("$NetBSD: strsep.c,v 1.7 1998/02/03 18:49:23 perry Exp $");
#endif
#endif /* LIBC_SCCS and not lint */

#include <string.h>

/*
 * Get next token from string *stringp, where tokens are possibly-empty
 * strings separated by characters from delim.  
 *
 * Writes NULs into the string at *stringp to end tokens.
 * delim need not remain constant from call to call.
 * On return, *stringp points past the last NUL written (if there might
 * be further tokens), or is NULL (if there are definitely no more tokens).
 *
 * If *stringp is NULL, strsep returns NULL.
 */
char *
strsep(stringp, delim)
	char **stringp;
	const char *delim;
{
	char *s;
	const char *spanp;
	int c, sc;
	char *tok;

	if ((s = *stringp) == NULL)
		return (NULL);
	for (tok = s;;) {
		c = *s++;
		spanp = delim;
		do {
			if ((sc = *spanp++) == c) {
				if (c == 0)
					s = NULL;
				else
					s[-1] = 0;
				*stringp = s;
				return (tok);
			}
		} while (sc != 0);
	}
	/* NOTREACHED */
}
